### ISSUE TITLE
Events bleed between tests using `run-test-async`

### ISSUE THREAD

**Issue by jasich** [Day8/re-frame-test#13]
I have a test where I want to check for one specific event using `rf-test/wait-for`.  This event cascades some dispatches of other events, which seem to bleed into other async tests if I don't explicitly wait for them to occur.  Ideally if the test ends re-frame-test would kill all pending events.  My current workaround is to add a `wait-for` at the end of the test for the last event in the event chain so that pending event's don't infect the other tests.

**Comment by danielneal** [comment]
I'm seeing this same problem. Specifically for me, I had some irrelevant dispatch-laters which were starting to bleed in to subsequent tests. I was only able to get rid of them by the following ugliness...

I'm working around it with the following horror

```clojurescript
(def *timers (atom #{}))

(defn monkey-patch-reframe-timers []
  (set! re-frame.interop/set-timeout!
        (fn [f ms]
          (let [id (js/setTimeout f ms)]
            (swap! *timers conj id)
            id))))

(defn clear-timers! []
  (doseq [t @*timers]
    (js/clearTimeout t))
  (reset! *timers #{}))
```

And calling clear-timers! as part of test setup. 

An official solution would be very, very welcome!!

**Comment by mike-thompson-day8** [comment]
Checking:  is the solution that `run-test-async`  clears all events from re-frame's queue when it exits?

**Comment by danielneal** [comment]
@mike-thompson-day8 definitely something like that. 
Would clearing re-frame's queue clear up the `dispatch-laters` though?
https://github.com/Day8/re-frame/blob/cac2c7a02e6680f4bf43d9b3d3cf72a2a39a3209/src/re_frame/fx.cljc#L88-L94 Looks like it boils down to`js/setTimeout`, I couldn't make out if there is a reframe layer queue of these that can be cleared, or if those timeouts just kinda hang around.

**Comment by mike-thompson-day8** [comment]
@danielneal  Sorry, I edited/added-to my last answer at about the same time you responded.  So you might not have seen everything I ended up putting in there.  Can you have another read please.

I see two different issues (solved in different ways):
  1.  Clearing the re-frame event queue 
  2.  Stopping  :dispatch-later events from ever happening.

**Comment by danielneal** [comment]
@mike-thompson-day8 Yes that sounds right! A stub of dispatch later (and any user defined fx that might leak via timeouts) would handle the slow bleed, and I clearing the event queue should handle the fast bleed of events that are already queued up.

**Comment by danielneal** [comment]
@mike-thompson-day8 What do I call to clear the event queue? Is that a change to re-frame-test or something I should be doing from my test code.

**Comment by mike-thompson-day8** [comment]
@danielneal  ` re-frame` doesn't currently supply API for purging events.  It probably should. We can add that.  https://github.com/Day8/re-frame/issues/425

Once that's in place, then the macro `run-test-async` can be changed to call this API

**Comment by mike-thompson-day8** [comment]
As a first step, I've hastily added `re-frame.core/purge-event-queue`.  Current code in the `re-frame` repo is untested.  That will happen tomorrow.  But there is enough there for someone to create a PR for the necessary, further changes to `run-test-async`

**Comment by jasich** [comment]
@mike-thompson-day8, I believe that clearing the queue at the end of `run-test-async` should fix the issue I was having.  I wonder if you'd want to communicate that the queue is clearing {x} number of events at the end of an async test though.  It might be good information when you're trying to debug a test issue. Also thanks for your work on re-frame & re-frame-test!

**Comment by spacegangster** [comment]
Hello @mike-thompson-day8 and the maintainers. Firstly, thank you for your awesome work!
If I could send a couple of bucks a month to fund further work – I would gladly do that :)

Official fix is still relevant. My run-test-async don't finish, until I exclude tests for the code that uses dispatch later. 

I'm using slightly modified code from @danielneal, and tests seem to work. (They work even with the original piece).

```
(def *timers (atom #{}))

(defn monkey-patch-reframe-timers! []
  (rf/reg-event-fx  :dispatch-later  (fn [_]))
  (set! re-frame.interop/set-timeout!
        (fn [f ms]
          (let [id (js/setTimeout f ms)]
            (swap! *timers conj id)
            id))))

(defn clear-timers! []
  (rf/purge-event-queue)
  (doseq [t @*timers]
    (js/clearTimeout t))
  (reset! *timers #{}))
```

If you need more input we can schedule a call.
If this workaround is good enough – I can incorporate it, run tests and do a PR.

Thanks again for the tools!
